File: cpp/src/patterns.cpp
#include "camintrinsics/patterns.hpp"

#include <opencv2/calib3d.hpp>
#include <opencv2/imgproc.hpp>

#include <cmath>

namespace ci {
namespace {
const cv::Scalar kBg(26, 26, 30);
}

cv::Mat gridImage(int width, int height, int step) {
  cv::Mat img(height, width, CV_8UC3, kBg);
  const cv::Scalar color(200, 200, 200), accent(60, 170, 250);
  for (int x = 0; x <= width; x += step) {
    const bool major = (x / step) % 5 == 0;
    cv::line(img, {x, 0}, {x, height}, major ? accent : color, major ? 2 : 1,
             cv::LINE_AA);
  }
  for (int y = 0; y <= height; y += step) {
    const bool major = (y / step) % 5 == 0;
    cv::line(img, {0, y}, {width, y}, major ? accent : color, major ? 2 : 1,
             cv::LINE_AA);
  }
  cv::rectangle(img, {1, 1}, {width - 2, height - 2}, cv::Scalar(90, 230, 90), 2);
  return img;
}

cv::Mat checkerboardImage(int width, int height, int squares, int margin) {
  cv::Mat img(height, width, CV_8UC3, kBg);
  const int side = std::min(width - 2 * margin, height - 2 * margin) / squares;
  const int x0 = (width - side * squares) / 2, y0 = (height - side * squares) / 2;
  for (int r = 0; r < squares; ++r)
    for (int c = 0; c < squares; ++c)
      if ((r + c) % 2 != 0)
        cv::rectangle(img, {x0 + c * side, y0 + r * side},
                      {x0 + (c + 1) * side, y0 + (r + 1) * side},
                      cv::Scalar(235, 235, 235), -1);
  cv::rectangle(img, {x0, y0}, {x0 + side * squares, y0 + side * squares},
                cv::Scalar(60, 170, 250), 2);
  return img;
}

cv::Mat radialTarget(int width, int height, int rings, int spokes) {
  cv::Mat img(height, width, CV_8UC3, kBg);
  const double cx = width / 2.0, cy = height / 2.0;
  const double rmax = std::hypot(cx, cy);
  for (int i = 1; i <= rings; ++i)
    cv::circle(img, {int(cx), int(cy)}, int(rmax * i / rings),
               cv::Scalar(200, 200, 200), 1, cv::LINE_AA);
  for (int k = 0; k < spokes; ++k) {
    const double a = 2 * CV_PI * k / spokes;
    cv::line(img, {int(cx), int(cy)},
             {int(cx + rmax * std::cos(a)), int(cy + rmax * std::sin(a))},
             cv::Scalar(110, 110, 110), 1, cv::LINE_AA);
  }
  cv::drawMarker(img, {int(cx), int(cy)}, cv::Scalar(60, 60, 235),
                 cv::MARKER_CROSS, 20, 2);
  return img;
}

cv::Mat photoLike(int width, int height) {
  cv::Mat img(height, width, CV_8UC3, cv::Scalar(60, 45, 35));
  const int horizon = int(height * 0.45);
  cv::rectangle(img, {0, 0}, {width, horizon}, cv::Scalar(150, 110, 70), -1);
  for (int i = -4; i <= 4; ++i)
    cv::line(img, {width / 2, horizon}, {width / 2 + i * width / 9, height},
             cv::Scalar(230, 230, 230), 3, cv::LINE_AA);
  const int bx[4] = {40, 200, 520, 690}, bw[4] = {120, 90, 140, 90};
  const double bh[4] = {0.30, 0.22, 0.34, 0.20};
  for (int b = 0; b < 4; ++b) {
    const int top = int(horizon - height * bh[b]);
    cv::rectangle(img, {bx[b], top}, {bx[b] + bw[b], horizon},
                  cv::Scalar(95, 95, 105), -1);
    cv::rectangle(img, {bx[b], top}, {bx[b] + bw[b], horizon},
                  cv::Scalar(40, 40, 45), 2);
    for (int wy = top + 12; wy < horizon - 10; wy += 26)
      for (int wx = bx[b] + 10; wx < bx[b] + bw[b] - 14; wx += 24)
        cv::rectangle(img, {wx, wy}, {wx + 12, wy + 14},
                      cv::Scalar(70, 190, 230), -1);
  }
  cv::line(img, {0, horizon}, {width, horizon}, cv::Scalar(230, 230, 230), 2,
           cv::LINE_AA);
  return img;
}

cv::Mat distortImage(const cv::Mat& ideal, const Mat33& K, const Dist& D,
                     const Mat33& KIdeal) {
  const int ow = ideal.cols, oh = ideal.rows;
  const double rLimit = maxDistortedRadius(D);
  cv::Mat mx(oh, ow, CV_32FC1), my(oh, ow, CV_32FC1);
  for (int v = 0; v < oh; ++v) {
    float* rowx = mx.ptr<float>(v);
    float* rowy = my.ptr<float>(v);
    for (int u = 0; u < ow; ++u) {
      const cv::Point2d xyd = pixelToNormalized({double(u), double(v)}, K);
      const cv::Point2d xy = undistortNormalized(xyd, D, rLimit);
      if (!std::isfinite(xy.x) || !std::isfinite(xy.y)) {
        rowx[u] = -1.0f;      // no ideal point maps here; remap -> borderValue
        rowy[u] = -1.0f;
        continue;
      }
      const cv::Point2d src = normalizedToPixel(xy, KIdeal);
      rowx[u] = float(src.x);
      rowy[u] = float(src.y);
    }
  }
  cv::Mat out;
  cv::remap(ideal, out, mx, my, cv::INTER_LINEAR, cv::BORDER_CONSTANT,
            cv::Scalar(12, 12, 14));
  return out;
}

cv::Mat undistortImage(const cv::Mat& distorted, const Mat33& K, const Dist& D,
                       double alpha, Mat33* KNew, cv::Rect* roi) {
  cv::Rect r;
  const Mat33 Kn = optimalNewCameraMatrix(K, D, distorted.cols, distorted.rows,
                                          alpha, &r);
  cv::Mat out;
  // cv::undistort itself is exact -- it maps destination to source with the
  // *forward* model. Only the new camera matrix above needed replacing.
  cv::undistort(distorted, out, cv::Mat(K), cv::Mat(D), cv::Mat(Kn));
  if (KNew) *KNew = Kn;
  if (roi) *roi = r;
  return out;
}

void distortionField(const Mat33& K, const Dist& D, int width, int height,
                     int step, std::vector<cv::Point2d>& uv,
                     std::vector<cv::Point2d>& delta) {
  uv.clear();
  delta.clear();
  for (int v = step / 2; v < height; v += step) {
    for (int u = step / 2; u < width; u += step) {
      const cv::Point2d p(u, v);
      const cv::Point2d xy = pixelToNormalized(p, K);  // treat as an ideal ray
      const cv::Point2d pd = normalizedToPixel(distortNormalized(xy, D), K);
      uv.push_back(p);
      delta.push_back(pd - p);
    }
  }
}

}  // namespace ci

File: python/camintrinsics/patterns.py
"""Synthetic 2D test images plus the two image warps that D controls.

The images here are *ideal pinhole* images: they are what a perfect, perfectly
undistorted camera with intrinsics ``K_ideal`` would record.  Feeding them
through :func:`distort_image` produces what a real lens would have delivered,
which lets you see distortion without owning the lens.
"""

from __future__ import annotations

import cv2
import numpy as np

from .intrinsics import (make_K, normalized_to_pixel,
                         optimal_new_camera_matrix, pixel_to_normalized,
                         undistort_normalized)

__all__ = ["grid_image", "checkerboard_image", "radial_target", "photo_like",
           "distort_image", "undistort_image", "distortion_field"]

BG = (26, 26, 30)


# --------------------------------------------------------------------------
# Test charts
# --------------------------------------------------------------------------
def grid_image(width: int = 800, height: int = 600, step: int = 40,
               color=(200, 200, 200), accent=(60, 170, 250)) -> np.ndarray:
    """A ruled grid: the reference chart for spotting bent straight lines."""
    img = np.full((height, width, 3), BG, np.uint8)
    for x in range(0, width + 1, step):
        c = accent if (x // step) % 5 == 0 else color
        cv2.line(img, (x, 0), (x, height), c, 1 if c is color else 2, cv2.LINE_AA)
    for y in range(0, height + 1, step):
        c = accent if (y // step) % 5 == 0 else color
        cv2.line(img, (0, y), (width, y), c, 1 if c is color else 2, cv2.LINE_AA)
    cv2.rectangle(img, (1, 1), (width - 2, height - 2), (90, 230, 90), 2)
    return img


def checkerboard_image(width: int = 800, height: int = 600, squares: int = 10,
                       margin: int = 40) -> np.ndarray:
    """A checkerboard plate, the pattern real calibrations are built on."""
    img = np.full((height, width, 3), BG, np.uint8)
    side = min(width - 2 * margin, height - 2 * margin) // squares
    x0 = (width - side * squares) // 2
    y0 = (height - side * squares) // 2
    for r in range(squares):
        for c in range(squares):
            if (r + c) % 2 == 0:
                continue
            cv2.rectangle(img, (x0 + c * side, y0 + r * side),
                          (x0 + (c + 1) * side, y0 + (r + 1) * side),
                          (235, 235, 235), -1)
    cv2.rectangle(img, (x0, y0), (x0 + side * squares, y0 + side * squares),
                  (60, 170, 250), 2)
    return img


def radial_target(width: int = 800, height: int = 600, rings: int = 10,
                  spokes: int = 24) -> np.ndarray:
    """Concentric circles + radial spokes, centred on the image centre.

    Radial distortion moves points *along* the spokes and leaves their angle
    untouched, so this chart separates the radial terms (circles change size)
    from the tangential ones (the whole pattern goes lopsided).
    """
    img = np.full((height, width, 3), BG, np.uint8)
    cx, cy = width / 2.0, height / 2.0
    rmax = np.hypot(cx, cy)
    for i in range(1, rings + 1):
        r = rmax * i / rings
        cv2.circle(img, (int(cx), int(cy)), int(r), (200, 200, 200), 1, cv2.LINE_AA)
    for k in range(spokes):
        a = 2 * np.pi * k / spokes
        cv2.line(img, (int(cx), int(cy)),
                 (int(cx + rmax * np.cos(a)), int(cy + rmax * np.sin(a))),
                 (110, 110, 110), 1, cv2.LINE_AA)
    cv2.drawMarker(img, (int(cx), int(cy)), (60, 60, 235), cv2.MARKER_CROSS, 20, 2)
    return img


def photo_like(width: int = 800, height: int = 600) -> np.ndarray:
    """A cartoon 'street scene': straight edges everywhere, like a real road."""
    img = np.full((height, width, 3), (60, 45, 35), np.uint8)
    cv2.rectangle(img, (0, 0), (width, int(height * 0.45)), (150, 110, 70), -1)
    for i in range(-4, 5):                                   # lane markings
        x = width // 2 + i * width // 9
        cv2.line(img, (width // 2, int(height * 0.45)), (x, height),
                 (230, 230, 230), 3, cv2.LINE_AA)
    for bx, bw, bh in [(40, 120, 0.30), (200, 90, 0.22), (520, 140, 0.34),
                       (690, 90, 0.20)]:                     # buildings
        top = int(height * 0.45 - height * bh)
        cv2.rectangle(img, (bx, top), (bx + bw, int(height * 0.45)), (95, 95, 105), -1)
        cv2.rectangle(img, (bx, top), (bx + bw, int(height * 0.45)), (40, 40, 45), 2)
        for wy in range(top + 12, int(height * 0.45) - 10, 26):
            for wx in range(bx + 10, bx + bw - 14, 24):
                cv2.rectangle(img, (wx, wy), (wx + 12, wy + 14), (70, 190, 230), -1)
    cv2.line(img, (0, int(height * 0.45)), (width, int(height * 0.45)),
             (230, 230, 230), 2, cv2.LINE_AA)
    return img


# --------------------------------------------------------------------------
# The two warps
# --------------------------------------------------------------------------
def distort_image(ideal: np.ndarray, K: np.ndarray, D, K_ideal=None,
                  size=None) -> np.ndarray:
    """Render what a lens with ``(K, D)`` would have recorded.

    For every pixel of the **output** (the real, distorted image) we ask "which
    ideal pixel does this see?":

    1. ``K^-1``  : output pixel  -> *distorted* normalized coords
    2. undistort : -> ideal normalized coords
    3. ``K_ideal``: -> a location in the ideal source image
    4. ``remap`` : sample the source there

    Note that the map is built **backwards**, from destination to source.  Every
    image warp works this way: forward-mapping leaves holes, backward-mapping
    does not.

    Output pixels whose radius lies past the distortion model's fold point are
    left at the border colour: no ideal point maps there, so any value would be
    a fabrication.  ``cv2.undistort`` fabricates one instead, which is where the
    swirling "petals" at the edge of an over-distorted image come from.
    """
    h, w = ideal.shape[:2]
    if K_ideal is None:
        K_ideal = make_K(w / 2.0, w / 2.0, w / 2.0, h / 2.0)
    ow, oh = size if size is not None else (w, h)
    vv, uu = np.mgrid[0:oh, 0:ow].astype(np.float64)
    uv = np.stack([uu.ravel(), vv.ravel()], axis=1)
    xy = undistort_normalized(pixel_to_normalized(uv, K), D)
    src = normalized_to_pixel(xy, K_ideal)
    # undistort_normalized returns NaN where no ideal point maps; -1 sends
    # remap to its border colour instead of fabricating a pixel.
    src[~np.isfinite(src)] = -1.0
    mx = src[:, 0].reshape(oh, ow).astype(np.float32)
    my = src[:, 1].reshape(oh, ow).astype(np.float32)
    return cv2.remap(ideal, mx, my, cv2.INTER_LINEAR,
                     borderMode=cv2.BORDER_CONSTANT, borderValue=(12, 12, 14))


def undistort_image(distorted: np.ndarray, K: np.ndarray, D, alpha: float = 0.0,
                    size=None):
    """Straighten an image, returning ``(image, K_new, roi)``.

    ``alpha`` is the trade-off knob of :func:`cv2.getOptimalNewCameraMatrix`:

    * ``alpha = 0`` -- zoom in until no invalid (black) pixel remains.  You lose
      field of view.
    * ``alpha = 1`` -- keep every original pixel.  You keep the FOV but gain
      black curved borders, and ``K_new`` has a smaller focal length.

    Either way **K changes**, and using the old K on the undistorted image is a
    silent, very common source of reprojection error.
    """
    h, w = distorted.shape[:2]
    ow, oh = size if size is not None else (w, h)
    K_new, roi = optimal_new_camera_matrix(K, D, w, h, float(alpha))
    # cv2.undistort itself is exact -- it maps destination to source with the
    # *forward* model. Only the K_new above needed replacing.
    out = cv2.undistort(distorted, K, np.asarray(D, float), None, K_new)
    return out, K_new, roi


def distortion_field(K: np.ndarray, D, width: int, height: int, step: int = 40):
    """Per-pixel displacement introduced by D, as ``(points, deltas)`` arrays.

    ``delta = distorted_pixel - ideal_pixel`` at the same normalized ray, i.e.
    "how far did the lens push this point, in pixels".
    """
    vv, uu = np.mgrid[step // 2:height:step, step // 2:width:step]
    uv = np.stack([uu.ravel().astype(float), vv.ravel().astype(float)], axis=1)
    xy = pixel_to_normalized(uv, K)              # treat as ideal rays
    from .intrinsics import distort_normalized
    uv_d = normalized_to_pixel(distort_normalized(xy, D), K)
    return uv, uv_d - uv
